pkg/importer/registry_data_source.go

```diff
@@ -0,0 +1,146 @@
+limitations under the License.
+*/
+
+package importer
+
+import (
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/klog"
+
+	"kubevirt.io/containerized-data-importer/pkg/image"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+const (
+	//containerDiskImageDir - Expected disk image location in container image as described in
+	//https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md
+	containerDiskImageDir = "disk"
+)
+
+// RegistryDataSource is the struct containing the information needed to import from a registry data source.
+// Sequence of phases:
+// 1. Info -> Transfer
+// 2. Transfer -> Process
+// 3. Process -> Convert (In the process phase the container image layers are extracted, and the url is pointed to the file determined to be the disk image)
+type RegistryDataSource struct {
+	endpoint    string
+	accessKey   string
+	secKey      string
+	certDir     string
+	insecureTLS bool
+	imageDir    string
+	//The discovered image file in scratch space.
+	url *url.URL
+}
+
+// NewRegistryDataSource creates a new instance of the Registry Data Source.
+func NewRegistryDataSource(endpoint, accessKey, secKey, certDir string, insecureTLS bool) *RegistryDataSource {
+	return &RegistryDataSource{
+		endpoint:    endpoint,
+		accessKey:   accessKey,
+		secKey:      secKey,
+		certDir:     certDir,
+		insecureTLS: insecureTLS,
+	}
+}
+
+// Info is called to get initial information about the data. No information available for registry currently.
+func (rd *RegistryDataSource) Info() (ProcessingPhase, error) {
+	return Transfer, nil
+}
+
+// Transfer is called to transfer the data from the source registry to a temporary location.
+func (rd *RegistryDataSource) Transfer(path string) (ProcessingPhase, error) {
+	if util.GetAvailableSpace(path) <= int64(0) {
+		// No scratch space available, exit with code indicating we need scratch space.
+		return Error, ErrRequiresScratchSpace
+	}
+	rd.imageDir = filepath.Join(path, containerDiskImageDir)
+
+	klog.V(1).Infof("Copying registry image to scratch space.")
+	err := image.CopyRegistryImage(rd.endpoint, path, containerDiskImageDir, rd.accessKey, rd.secKey, rd.certDir, rd.insecureTLS)
+	if err != nil {
+		return Error, errors.Wrapf(err, "Failed to read registry image")
+	}
+
+	return Process, nil
+}
+
+// Process is called to do any special processing before giving the url to the data back to the processor
+func (rd *RegistryDataSource) Process() (ProcessingPhase, error) {
+	imageFile, err := getImageFileName(rd.imageDir)
+	if err != nil {
+		return Error, errors.Wrapf(err, "Cannot locate image file")
+	}
+
+	// imageFile and rd.imageDir are both valid, thus the Join will be valid, and the parse will work, no need to check for parse errors
+	rd.url, _ = url.Parse(filepath.Join(rd.imageDir, imageFile))
+	klog.V(3).Infof("Successfully found file. VM disk image filename is %s", rd.url.String())
+	return Convert, nil
```

Comment:
According https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md, which is the format we support, the only format stored in a registry disk should be RAW (otherwise kubevirt can't boot it). So I specifically didn't run the registry import through the format readers. Technically we should fail the convert if we find a qcow2 disk image as that also doesn't conform to the format outlined in that document. But then we get back to the should we add more complex code (to detect qcow2 and reject it specifically) vs simpler code.